[Fix #52304] Avoid computing `klass` if reflection is a `belongs_to` in `ActiveRecord::AutosaveAssociation#inverse_belongs_to_association_for`

joshuay03 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

Fixes https://github.com/rails/rails/issues/52304
Ref: https://github.com/rails/rails/pull/49847

### Detail

Adds a condition in `ActiveRecord::AutosaveAssociation#inverse_belongs_to_association_for` to return early if the current reflection is a `belongs_to`, as the inverse can't possibly be a `belongs_to` as well. This then ensures that polymorphic `belongs_to` reflections don't have their class computed which raises an error like the one in the linked issue (see https://github.com/rails/rails/pull/31895 for why this happens).

An alternative fix, specifically for the linked issue is to simply return if `reflection.polymorphic?`, but imo what I've currently got is better since it handles non-polymorphic cases as well, and will short circuit in all cases where the inverse logic is not applicable.

**Edit:** Some more context in https://github.com/rails/rails/pull/52305#issuecomment-2222227506

### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

joshuay03 committed:
[Fix #52304] Avoid computing `klass` if reflection is a `belongs_to` in `ActiveRecord::AutosaveAssociation#inverse_belongs_to_association_for`

ghiculescu commented:
I probably should have made a PR, but the tests in https://github.com/rails/rails/issues/52304 are copy-pasted from replication tests I wrote in https://github.com/rails/rails/blob/main/activerecord/test/cases/associations/has_one_through_associations_test.rb to confirm the issue.

So you'd need to add a `Certificate` model and a `Sponsor` model.

joshuay03 commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -314,8 +314,18 @@ def test_callbacks_on_child_when_parent_autosaves_child
   end
 
+  def test_callbacks_on_child_when_parent_autosaves_polymorphic_child_with_inverse_of
+    drink_designer = DrinkDesigner.create!(chef: Chef.new)
```
Without patch:
<img width="1728" alt="Screenshot 2024-07-11 at 12 49 24 pm" src="https://github.com/rails/rails/assets/54629302/064f7447-032e-416b-b903-8a0d47af9e5c">

joshuay03 commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -324,6 +334,16 @@ def test_callbacks_on_child_when_child_autosaves_parent
   end
 
+  def test_callbacks_on_child_when_polymorphic_child_with_inverse_of_autosaves_parent
+    chef = Chef.create!(employable: DrinkDesigner.new)
```
Without patch:
<img width="1728" alt="Screenshot 2024-07-11 at 12 49 40 pm" src="https://github.com/rails/rails/assets/54629302/f792c3a5-dbd7-4b90-b39e-6d0ce0137df8">

joshuay03 commented:
> I probably should have made a PR, but the tests in #52304 are copy-pasted from replication tests I wrote in [`main`/activerecord/test/cases/associations/has_one_through_associations_test.rb](https://github.com/rails/rails/blob/main/activerecord/test/cases/associations/has_one_through_associations_test.rb?rgh-link-date=2024-07-11T06%3A59%3A05Z) to confirm the issue.
> 
> So you'd need to add a `Certificate` model and a `Sponsor` model.

All good, I've added some failing tests similar to the ones I added in https://github.com/rails/rails/pull/49847.

It turns out this is only an issue if the validate association callbacks are run (either through `validate: true` like you did or autosave which I'm doing with `accepts_nested_attributes_for`) **AND** `:inverse_of` is set on the polymorphic `belongs_to` association.

Without the `inverse_of` the class seems to compute just fine. Might be something to look into separately but I'm gonna consider it out of scope for this since the condition I added is a good change anyway.

joshuay03 commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -314,8 +314,18 @@ def test_callbacks_on_child_when_parent_autosaves_child
   end
 
+  def test_callbacks_on_child_when_parent_autosaves_polymorphic_child_with_inverse_of
+    drink_designer = DrinkDesigner.create!(chef: ChefWithPolymorphicInverseOf.new)
```
Without patch:
<img width="1728" alt="Screenshot 2024-07-11 at 1 05 04 pm" src="https://github.com/rails/rails/assets/54629302/fb9c26e1-9595-4f15-82c6-cdbd944ab311">

joshuay03 commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -324,6 +334,16 @@ def test_callbacks_on_child_when_child_autosaves_parent
   end
 
+  def test_callbacks_on_child_when_polymorphic_child_with_inverse_of_autosaves_parent
+    chef = ChefWithPolymorphicInverseOf.create!(employable: DrinkDesigner.new)
```
Without patch:
<img width="1728" alt="Screenshot 2024-07-11 at 1 06 05 pm" src="https://github.com/rails/rails/assets/54629302/a96cf6c3-50d1-4875-8a02-e5b5fb89247d">
